[Merged by Bors] - feat(analysis/calculus/times_cont_diff, analysis/calculus/inverse): smooth inverse function theorem

hrmacbeth opened the pull request:
The inverse function theorem, in the C^k and smooth categories.

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

---

~~There are a couple of lemmas which seem like they should be much shorter if approached the right way, or refactored to extract the right simp-lemmas: `continuous_linear_equiv.is_open`, `image_to_local_homeomorph_image_basepoint`.  Perhaps @urkud or @sgouezel could take a look.~~ (Update: these lemmas have been cleaned up thanks to @urkud.)

- [x] depends on: #4511

hrmacbeth committed:
feat(analysis/calculus/times_cont_diff) smooth inv

hrmacbeth committed:
Merge branch 'master' into smooth-inverse

bryangingechen commented on src/analysis/calculus/fderiv.lean:
```diff
@@ -1914,6 +1914,61 @@ h.continuous.comp (continuous_const.prod_mk continuous_id)
 
+namespace continuous_linear_equiv
+
+/-! ### The set of continuous linear equivalences between two Banach spaces is open
```
```suggestion
/-!
### The set of continuous linear equivalences between two Banach spaces is open
```

bryangingechen commented on src/analysis/calculus/fderiv.lean:
```diff
@@ -1914,6 +1914,61 @@ h.continuous.comp (continuous_const.prod_mk continuous_id)
+begin
+  refine mem_nhds_sets continuous_linear_equiv.is_open _,
+  simp
+end
```
```suggestion
mem_nhds_sets continuous_linear_equiv.is_open (by simp)
```

hrmacbeth committed:
Apply suggestions from code review

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

hrmacbeth commented on src/analysis/calculus/inverse.lean:
```diff
@@ -496,3 +496,90 @@ theorem to_local_left_inverse {g : 𝕜 → 𝕜} (hg : ∀ᶠ x in 𝓝 a, g (f
+  have := hf.image_to_local_homeomorph_image_basepoint hf' hn,
+  apply times_cont_diff_at.of_local_homeomorph (image_mem_to_local_homeomorph_target hf hf' hn),
+  convert hf',
+  convert hf
```
```suggestion
  { convert hf' },
  { convert hf }
```

hrmacbeth committed:
Update src/analysis/calculus/inverse.lean

urkud commented on src/analysis/calculus/fderiv.lean:
```diff
@@ -1914,6 +1914,59 @@ h.continuous.comp (continuous_const.prod_mk continuous_id)
+of its differentiability.
+-/
+
+lemma is_open [complete_space E] :
```
```suggestion
protected lemma is_open [complete_space E] :
```

hrmacbeth replied:
Could you explain this change?  IIUC `protected` is for things that are not generally useful, but this fact does seem generally useful

bryangingechen replied:
> IIUC `protected` is for things that are not generally useful, but this fact does seem generally useful

I think you might be confusing `private` and `protected`. Briefly,
- `private` makes it so that you can't refer at all to a declaration (except by trickery) outside the section or file; I think we mostly mark things `private` in mathlib when they're followed by something strictly more general.
- `protected` makes it so that the declaration can only be referred to in fully-qualified form ([obligatory TPiL link](https://leanprover.github.io/theorem_proving_in_lean/interacting_with_lean.html#more-on-namespaces)). In this case, even if we `open continuous_linear_equiv`, we still have to refer to this lemma using `continuous_linear_equiv.is_open`. Marking things `protected` is useful when there are lots of things with the same name under different namespaces.

hrmacbeth replied:
Ah!  Thanks.

urkud committed:
Update src/analysis/calculus/fderiv.lean

hrmacbeth commented on src/analysis/calculus/inverse.lean:
```diff
@@ -496,3 +496,90 @@ theorem to_local_left_inverse {g : 𝕜 → 𝕜} (hg : ∀ᶠ x in 𝓝 a, g (f
+
+/-- Given a `times_cont_diff` function over `ℝ` with an invertible derivative at `a`, the inverse
+function (produced by `times_cont_diff.to_local_homeomorph`) is also `times_cont_diff`. -/
+lemma rename_me
```
and let's at least come up with a name for this lemma before merging!

urkud commented on src/analysis/calculus/inverse.lean:
```diff
@@ -496,3 +496,90 @@ theorem to_local_left_inverse {g : 𝕜 → 𝕜} (hg : ∀ᶠ x in 𝓝 a, g (f
+def to_local_homeomorph
+  {n : with_top ℕ} (hf : times_cont_diff_at ℝ n f a) (hf' : has_fderiv_at f (f' : E' →L[ℝ] F') a)
+  (hn : 1 ≤ n) :
+  local_homeomorph E' F' :=
```
If you prove
```lean
lemma times_cont_diff_at.has_strict_fderiv_at' (hf : times_cont_diff_at ℝ n f a) (hf' : has_fderiv_at f f' a) (hn : 1 ≤ n) :
  has_strict_fderiv_at f f' a :=
by simpa only [hf'.fderiv] using hf.has_strict_fderiv_at
```
then this can becone `(hf.has_strict_fderiv_at' hf' hn).to_local_homeomorph`. This should save you some `rw`s in the proofs. You can also define `times_cont_diff_at.local_inverse` as `(hf.has_strict_fderiv_at' hf' hn).local_inverse`, then lemma `rename_me` can be `times_cont_diff_at.to_local_inverse` mirroring `has_strict_fderiv_at` API.

hrmacbeth replied:
Great, thanks for this!

urkud commented on src/analysis/calculus/times_cont_diff.lean:
```diff
@@ -874,6 +874,16 @@ begin
 end
 
+lemma times_cont_diff_within_at_zero (hx : x ∈ s)
+  (h : ∃ u ∈ 𝓝[s] x, continuous_on f (s ∩ u)) :
```
Is it an `iff`?

hrmacbeth replied:
Yes, thanks! I made them simp lemmas too.

urkud commented on src/analysis/calculus/inverse.lean:
```diff
@@ -496,3 +496,90 @@ theorem to_local_left_inverse {g : 𝕜 → 𝕜} (hg : ∀ᶠ x in 𝓝 a, g (f
+  exact hf₂.image_mem_to_local_homeomorph_target
+end
+
+lemma image_to_local_homeomorph_image_basepoint
```
With the change proposed above you should be able to prove it using `(hf.has_strict_fderiv_at' hf' hn).local_inverse_apply_image`. Please sync names with `has_strict_fderiv_at`.

hrmacbeth committed:
new inverse-function API

hrmacbeth commented on src/analysis/calculus/fderiv.lean:
```diff
@@ -1914,6 +1914,59 @@ h.continuous.comp (continuous_const.prod_mk continuous_id)
+    refine ⟨λ h', _, λ h', empty_subset _ h'⟩,
+    obtain ⟨e, he⟩ := h',
+    exfalso,
+    exact not_nonempty_iff_imp_false.mp h e },
```
This (`continuous_linear_map.is_open`) is the remaining lemma whose proof I was unhappy with.  I was thinking of extracting the highlighted section as a lemma that in a not-nonempty topological space, every set is open.  (Maybe, that a not-nonempty topological space has discrete topology.). What do you think?  The other parts of the lemma might be golf-able.

urkud replied:
If you start with `is_open_iff_nhds`, then you have an element of your set immediately.

urkud replied:
I'm golfing this lemma.

urkud replied:
Done. I've reformulated this lemma using `range`. The new lemma is not defeq to the old one because of lhs/rhs in the equality. I hope, this won't cause too much trouble.

urkud replied:
It seems that the new `subsingleton` instances make a proof break. I'll fix it tonight.

urkud replied:
I've created #4511.

hrmacbeth replied:
Thanks for investigating this!

hrmacbeth committed:
authors and module doc

urkud committed:
Golf a proof

urkud committed:
More subsingleton/unique instances

urkud committed:
Merge branch 'master' into smooth-inverse

urkud committed:
chore(*): add a few more `unique` instances

* `linear_map.unique_of_left`, `linear_map.unique_of_right`,
  `continuous_linear_map.unique_of_left`,
  `continuous_linear_map.unique_of_right`: if either `M` or `M₂` is a
  `subsingleton`, then both `M →ₗ[R] M₂` and `M →L[R] M₂` are
  `unique`;
* `pi.unique`: if each `β a` is `unique`, then `Π a, β a` is `unique`;
* rename `function.injective.comap_subsingleton` to
  `function.injective.subsingleton`;
* add `unique.mk'` and `function.injective.unique`;
* add a few `simp` lemmas for `default`;
* drop `nonempty_unique_or_exists_ne` and `subsingleton_or_exists_ne`.

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Split long line

urkud committed:
Merge branch 'unique-instances' into smooth-inverse

urkud committed:
Merge branch 'master' into smooth-inverse

urkud committed:
Fix merge error

hrmacbeth commented on src/analysis/calculus/inverse.lean:
```diff
@@ -31,8 +31,15 @@ and prove two versions of the inverse function theorem:
+We also reformulate the theorems in terms of `times_cont_diff`, to give that `C^k` (respectively,
+smooth) inputs give `C^k` (smooth) inverses.  These versions require that continuous
+differentiability implying strict differentiability; this is false over a general field, true over
+`ℝ` (the setting in which it is implemented here), and true but (TODO) not yet implemented over `C`.
```
```suggestion
`ℝ` (the setting in which it is implemented here), and true but (TODO) not yet implemented over `ℂ`.
```

hrmacbeth committed:
Update src/analysis/calculus/inverse.lean

urkud committed:
Fix proof

urkud committed:
Merge branch 'master' into smooth-inverse

sgouezel commented on src/analysis/calculus/inverse.lean:
```diff
@@ -31,8 +31,15 @@ and prove two versions of the inverse function theorem:
+
+We also reformulate the theorems in terms of `times_cont_diff`, to give that `C^k` (respectively,
+smooth) inputs give `C^k` (smooth) inverses.  These versions require that continuous
+differentiability implying strict differentiability; this is false over a general field, true over
```
```suggestion
differentiability implies strict differentiability; this is false over a general field, true over
```

sgouezel commented on src/analysis/calculus/times_cont_diff.lean:
```diff
@@ -873,6 +873,22 @@ begin
   exact ⟨by rwa insert_eq_of_mem hx, λ x hx, by simp [ftaylor_series_within]⟩
 end
 
+@[simp] lemma times_cont_diff_within_at_zero (hx : x ∈ s) :
```
I don't think this should be a simp lemma, the left hand side looks simpler than the right hand side to me.

sgouezel commented on src/analysis/calculus/times_cont_diff.lean:
```diff
@@ -1315,6 +1331,10 @@ begin
   exact times_cont_diff_on_zero
 end
 
+@[simp] lemma times_cont_diff_at_zero :
```
ditto, I don't think it should be a simp lemma.

sgouezel commented on src/analysis/calculus/times_cont_diff.lean:
```diff
@@ -2421,6 +2501,14 @@ begin
 end
 
+/-- If a function is `C^n` with `1 ≤ n` around a point, and its derivative of `f` at that point is
+given to us as `f`, then `f'` is also a strict derivative. -/
```
```suggestion
/-- If a function is `C^n` with `1 ≤ n` around a point, and its derivative of at that point is
given to us as `f'`, then `f'` is also a strict derivative. -/
```

sgouezel reviewed (commented):
This looks good. 

There are two obvious refactors that are calling here:
* do not deduce the continuity of bounded bilinear maps from their differentiability, to be able to put several facts at more natural places
* Generalize real stuff to `is_R_or_C` fields (or another typeclass if necessary)

All this can wait for later PRs.

hrmacbeth committed:
Apply suggestions from code review

Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>

hrmacbeth committed:
remove `simp`

hrmacbeth committed:
fix build

hrmacbeth commented on src/analysis/calculus/fderiv.lean:
```diff
@@ -1914,6 +1914,47 @@ h.continuous.comp (continuous_const.prod_mk continuous_id)
+
+In this section we establish that the set of continuous linear equivalences between two Banach
+spaces is an open subset of the space of linear maps between them.  These facts are placed here
+because the proof uses `is_bounded_bilinear_map_comp.continuous`, proved just above as a consequence
```
```suggestion
because the proof uses `is_bounded_bilinear_map.continuous`, proved just above as a consequence
```

hrmacbeth committed:
Update src/analysis/calculus/fderiv.lean
